### DISCUSSION TITLE
Circuit question

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#10] [General]
hi, I'd like to give this a try but is this disclaimer still up to date or are there changes to the diagram?

![Screenshot from 2021-11-14 00-43-46](https://user-images.githubusercontent.com/305679/141662193-23768aa1-d9a2-44f8-b45c-4f9288c1830a.png)

**Comment by corrados** [comment]
I am no hardware design expert. So, yes, the disclaimer is still valid. All I can say is that it works well for me.

Please note that this schematic does not show you all the needed components nor the input ports to the ESP32 are correct. There is just one schematic given for any piezo trigger and one for the hi-hat controller. If you want to use, e.g., 6 pads with stereo piezo support, you need to build the circuit for the piezo input 12 times and use the correct ESP32 input ports as given in
https://github.com/corrados/edrumulus/blob/10fa7ffbeec366b2b21db8af47cd8eb771b7b096/edrumulus.ino#L60
and
https://github.com/corrados/edrumulus/blob/10fa7ffbeec366b2b21db8af47cd8eb771b7b096/edrumulus.ino#L61

Please also note that Edrumulus is still just experimental software. It works ok but there is still a lot of room for improvement. Also, I only support the pads given in the enum
https://github.com/corrados/edrumulus/blob/10fa7ffbeec366b2b21db8af47cd8eb771b7b096/edrumulus.h#L26
If you have self-made pads, Edrumulus might not perform well at all.

**Comment by thijstriemstra** [comment]
i would like to test with my roland cy-5, roland pd-5 and pd-125bk. looks like this project is worth a try so ordering components for this project today!

**Comment by corrados** [comment]
Do you want to use the ESP32 or the Teensy? I would recommend the Teensy, actually.

**Comment by thijstriemstra** [comment]
i have a esp32 devkit v4 i wanted to use but i can try teensy as well.

**Comment by thijstriemstra** [comment]
ps. do you have the source of that fritzing sketch and could you add it to the repository? and any advice on what type of piezo trigger/hhat control ports? A 1/4 stereo jack female pcb mount with 2 pins? or 3? I really like these: https://www.ebay.com/itm/324731292802

![s-l16dd00](https://user-images.githubusercontent.com/305679/143793630-f79cea03-f150-4b97-b2b4-6d9dce8b8eff.jpg)

Also, is it possible to trigger this in a DAW using MIDI?

![edrumulus](https://user-images.githubusercontent.com/305679/141684966-ae7c352d-8f08-4cca-95e4-0268bc2c47ac.png)

**Comment by corrados** [comment]
> do you have the source of that fritzing sketch and could you add it to the repository?

Here it is:
[edrumulus_new.zip](https://github.com/corrados/edrumulus/files/7534081/edrumulus_new.zip)

> and any advice on what type of piezo trigger/hhat control ports? A 1/4 stereo jack female pcb mount with 2 pins? or 3?

I am using normal stereo jack ports which are soldered to a small PCB as seen in the following picture: https://github.com/corrados/edrumulus/blob/main/algorithm/images/jackboard.jpg. These PCBs are taken from an old patchbay.

> I really like these: https://www.ebay.com/itm/324731292802

Wow, this looks really professional. Even Roland modules do only use the simply types.

> Also, is it possible to trigger this in a DAW using MIDI?

Edrumulus just sends out MIDI notes over serial to the host PC. If you are using Windows, you need the Hairless MIDI tool to convert it to a MIDI stream which a DAW can use.

In my [Linux script](https://github.com/corrados/edrumulus/blob/main/tools/run_edrumulus.sh), I am using the ttymidi tool and send the MIDI signal to [Drumgizmo](https://drumgizmo.org).

**Comment by bzass** [comment]
> If you have self-made pads, Edrumulus might not perform well at all.

Hi, I'm also super impressed by your work - thanks for sharing this project and your active contributions! I would like to give it a try with ESP32 and a custom pad based on the Open Edrums project by Ryo Kosaka. I know that those are not currently supported. Would it be possible for you one day to write or record a short tutorial about your methodology of tuning parameters to the other pads or at least a practical explanation of parameters to allow for manual tuning? That would be much appreciated and make it possible for a broader community to tune them to their pads. Thanks once again!

**Comment by corrados** [comment]
> Would it be possible for you one day to write or record a short tutorial about your methodology of tuning parameters to the other pads or at least a practical explanation of parameters to allow for manual tuning?

At present time it would be better that you record a wave file of your piezo trigger signal captured with a sampling rate of 8 kHz and hitting the pad according to the rules defined in https://github.com/corrados/edrumulus/tree/main/algorithm#test-signals. I could then add your wave file in the "signals" directory of the Git repo: https://github.com/corrados/edrumulus/tree/main/algorithm/signals. I could then tune the parameters for you.

> That would be much appreciated and make it possible for a broader community to tune them to their pads.

The problem is that the algorithms are not yet fix. I want to improve the positional sensing and also want to tune some other parts of the algorithm. Each change of the algorithm will most probably require the pad parameters to be adjusted. Therefore it does not make sense to fine-tune the parameters yet.

**Comment by thijstriemstra** [comment]
Thanks for the feedback!

>  I would recommend the Teensy, actually.

Any reason? Faster? Cheaper?

> Wow, this looks really professional. Even Roland modules do only use the simply types.

Yea! Found [6 for $14](https://www.ebay.com/itm/124892424978) that I'll give a try.

**Comment by corrados** [comment]
> Any reason? Faster? Cheaper?

The ESP32 has some unsolved hardware issues. The ADC multiplexer seems not to separate the inputs good enough. E.g., if I move the Hi-Hat pedal it changes the DC offset of another piezo trigger input which is bad. I did not see such issues with the Teensy. Also, the ADC spikes are much smaller for the Teensy.

**Comment by bzass** [comment]
> The problem is that the algorithms are not yet fix. I want to improve the positional sensing and also want to tune some other parts of the algorithm. Each change of the algorithm will most probably require the pad parameters to be adjusted. Therefore it does not make sense to fine-tune the parameters yet.

Thanks a lot for your reply. Fully understood, I will wait for things to get more stable. It would be great to get some docs/tutorials after it happens so that we don't need to bother you with parameters for custom pads.

Thanks again for this great contribution.

**Comment by thijstriemstra** [comment]
> I did not see such issues with the Teensy.

thanks, didn't know that. I'll get a teensy 4.0.
 
![teensy](https://user-images.githubusercontent.com/305679/141693162-88a18d16-0451-4024-89aa-6b0bf82edaca.png)

**Reply by jstma** [reply]
@thijstriemstra The teensy 4.0 has a serious design defect that severely limits the adc performance.  I don't recommend the teensy 4.0.
Teensy 3.6 looks like a better choice if you can find one.

**Reply by corrados** [reply]
Do you think Edrumulus will work with the Teensy 3.6?

**Reply by thijstriemstra** [reply]
Thanks for the feedback @jstma. What about the newer teensy 4.1?

**Reply by jstma** [reply]
The 4.1 and 4.0 share the same design.  I actually have a 4.1 board and the adc is noisy.

**Reply by jstma** [reply]
@corrados  The Teensy 3.6 has a better design on paper, but it's been out of stock for awhile.  I haven't been able to find one for a resonable price.

**Reply by thijstriemstra** [reply]
found some reports about ADC speed [here](https://forum.pjrc.com/threads/57683-Analog-read-on-Teensy-4-0-slower-(compared-to-Teensy-3-6)) that confirm teensy 3.6 design is better. meh.

>  modified JBeale's code to separate the AnalogRead from the double precision floating point calculations. I connected the A0 to an opAmp buffer to get a nice stable voltage.

> Here's what I got:
> Teensy 3.6 - Samples/sec: 384568.03 Process Time (us): 392176 P-P noise: 11 St.Dev: 0.842
> Teensy 4.1 - Samples/sec: 176467.69 Process Time (us): 6667 P-P noise: 32 St.Dev: 1.563

> As you can see, the Teensy reads more than twice as fast, and gives much more accurate readings. In terms of double precision floating point math, the Teensy 4.1 slaughters in the T3.6. I will repeat the test with single precision floats. I suspect the difference in the Process time should all but disappear.

> And here are the results when 'double' is changed to 'float'
> T3.6 Samples/sec: 384568.03 Process Time (us): 11959 P-P noise: 10 St.Dev: 0.759
> T4.1 Samples/sec: 176675.33 Process Time (us): 4333 P-P noise: 35 St.Dev: 1.663

> Look like T4.1 can still process floats almost 3 times as fast at T3.6, which like reflects the mhz bump.

**Reply by jstma** [reply]
Speed is not the problem.  For more background you can read this [discussion](https://github.com/corrados/edrumulus/discussions/8).
It's not the rms noise that's the problem.  It's the spurious noise due to, I believe, a poor schematic/layout.

**Reply by thijstriemstra** [reply]
> I haven't been able to find one for a resonable price.

I'm trying to keep the cost below an eDrumin module, exc labor/weed, i found one locally for [35 euro](https://www.okaphone.com/artikel.asp?id=484498) which seems doable (for now). Ordered it.

**Reply by jstma** [reply]
The eDRUMin module is really good bang for the buck.  If it was open source I'd be all over it.

With a Teensy 3.6 you should be able to get 20 inputs just like the eDRUMin 10.  The only issue is the unkown cpu power required for edrumulus.

I've spent the last month or so looking for a better solution using discrete adcs or other mpus (it's time consuming) and the Teensy 3.6 is still at the top of my list.  It's just not available to everyone right now.  The only other mpu I've looked at that would be close is the STM32F373, but it's not readily available in a teensy/arduino like board and it doesn't have the horsepower to run edrumulus.  But in that case I would use the STM32F373 to read the piezos and send the data back to an ESP32 over spi.

I've actually been contemplating spinning a pcb using the STM32F373CCT6 which is available with a low minimum order (10 pcs min) on JLPCB, but that chip is out of stock *everywhere* so I doubt JLPCB could procure some.

In any case, the Teensy 3.6 is untested.  The adc looks okay, but I can't guarantee it.  You'll need to do some testing.

**Comment by corrados** [comment]
I have moved this discussion to the discussion forum.

**Comment by corrados** [comment]
> Fully understood, I will wait for things to get more stable.

On a Git side branch I am currently re-writing large portions of the Edrumulus signal processing code. I'll remove the Hilbert filter and use the scan/mask times in a more "standardized way", i.e., the scan time starts from the point the signal crosses the threshold (previously, it was started from the first main peak) and the mask time starts at the and of the scan time (previously, was also started from the first main peak).

The first part of this change is done now, i.e., the rework of the "drumtrigger.m" file. Now, I have to port this new code to edrumulus.m and the micro processor C++ code.

> It would be great to get some docs/tutorials after it happens so that we don't need to bother you with parameters for custom pads.

This is definitely on my list.

**Comment by thijstriemstra** [comment]
>  All I can say is that it works well for me.

@corrados maybe the disclaimer should reflect this.

**Reply by corrados** [reply]
[done](https://github.com/corrados/edrumulus/commit/a4cff5e230b421d4abeb9b016747d7876f81825c)

**Comment by thijstriemstra** [comment]
This might be a silly question, but are there any requirements for amount of ampere for my power supply? I have a 5v 1.0A laying around I'd like to use.

**Reply by jstma** [reply]
If you have it lying around, it can't hurt to try it.  I think 5V 1.0A is fine for this application, but I haven't tested it.

**Comment by thijstriemstra** [comment]
> Edrumulus just sends out MIDI notes over serial to the host PC. If you are using Windows, you need the Hairless MIDI tool to convert it to a MIDI stream which a DAW can use.

I see Teensy can be used as a MIDI USB device: https://www.pjrc.com/teensy/td_midi.html

> Teensyduino implements a "class compliant" MIDI device, which can work with the built-in drivers on all major operating systems. When MIDIx4 or MIDIx16 are selected, Teensy implements a class compliant device with 4 or 16 virtual cables/ports. 

Is this something that could be supported or does it introduce (too much) latency etc?

**Reply by corrados** [reply]
>> Edrumulus just sends out MIDI notes over serial to the host PC. If you are using Windows, you need the Hairless MIDI tool to convert it to a MIDI stream which a DAW can use.

> I see Teensy can be used as a MIDI USB device

Correct. You only need the Hairless MIDI tool for the ESP32

**Reply by thijstriemstra** [reply]
I haven't used Hairless MIDI before, and I'm probably missing something, but this is my perception of the current flow:

edrumulus teensy -> teensy serial over usb -> hairless midi on windows -> hairless midi device within ableton -> vst

but I was hoping something like this:

edrumulus teensy -> teensy midi over usb -> edrumulus midi device within ableton -> vst

so basically skipping the hairless thing.

**Reply by corrados** [reply]
> but I was hoping something like this:
> edrumulus teensy -> teensy midi over usb -> edrumulus midi device within ableton -> vst
> so basically skipping the hairless thing.

Yes, this is correct. The Teensy shows up as a MIDI device which can directly be used in your DAW.

**Reply by thijstriemstra** [reply]
awesome!! So basically Teensy supports MIDI USB out of the box and ESP32 needs Hairless.

**Comment by thijstriemstra** [comment]
@corrados any idea what 1, 2, and 3 represent on these plugs? More pictures on https://www.ebay.com/itm/124892424978

![s-l1600](https://user-images.githubusercontent.com/305679/145896372-13fe8c67-d779-41c8-a5ec-00bea7e2738f.jpg)

How does it tie into your circuit?

**Reply by bzass** [reply]
This is sleeve, right and left. Sleeve = GND (common for all piezos), right / left is for 2 piezos input. If you use one piezo, you just need one of those two like in example schema. I use other type of sockets where you can easily see which pin is which:

![image](https://user-images.githubusercontent.com/12543258/145905895-a9c33595-2bcd-4da2-badb-fe6206313e90.png)

Most likely 2 is for GND, 1 and 3 for tip / ring. You may verify it using multimeter (with beeper).

**Reply by thijstriemstra** [reply]
thanks for the feedback @bzass 

I don't have a beeping multimeter. For newbies, inc myself, I made this image, but is it correct..?

![hihat circuit](https://user-images.githubusercontent.com/305679/146688869-96929e91-2f7f-4c34-8733-06fdb3a8d196.jpg)

**Reply by bzass** [reply]
It's hard to say as you can't see the wiring in this socket but if 2 is GND as I assumed, 2 should be connected to sleeve (not right as in your picture). Check step 4 from here: https://create.arduino.cc/projecthub/ryokosaka/minimal-midi-drum-kit-with-3d-printer-5633d2

![image](https://user-images.githubusercontent.com/12543258/145910520-1e6b1954-f5ef-434e-99be-f7c849e69fa9.png)

**Reply by thijstriemstra** [reply]
Turns out `3` is left and `2` is sleeve. I updated the image in previous comment.
